src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Converters.cs

```diff
@@ -70,7 +70,72 @@ public sealed partial class JsonSerializerOptions
             return converters;
 
             void Add(JsonConverter converter) =>
-                converters.Add(converter.TypeToConvert!, converter);
+                converters.Add(converter.TypeToConvert, converter);
+        }
+
+        internal JsonConverter GetDictionaryKeyConverter(Type keyType)
+        {
+            _dictionaryKeyConverters ??= GetDictionaryKeyConverters();
+
+            if (!_dictionaryKeyConverters.TryGetValue(keyType, out JsonConverter? converter))
+            {
+                if (keyType.IsEnum)
+                {
+                    converter = GetEnumConverter();
+                    _dictionaryKeyConverters[keyType] = converter;
+                }
+                else
+                {
+                    ThrowHelper.ThrowNotSupportedException_DictionaryKeyTypeNotSupported(keyType);
+                }
+            }
+
+            return converter!;
+
+            // Use factory pattern to generate an EnumConverter with AllowStrings and AllowNumbers options for dictionary keys.
+            // There will be one converter created for each enum type.
+            JsonConverter GetEnumConverter()
+                => (JsonConverter)Activator.CreateInstance(
+                        typeof(EnumConverter<>).MakeGenericType(keyType),
+                        BindingFlags.Instance | BindingFlags.Public,
+                        binder: null,
+                        new object[] { EnumConverterOptions.AllowStrings | EnumConverterOptions.AllowNumbers, this },
+                        culture: null)!;
+        }
+
+        private ConcurrentDictionary<Type, JsonConverter>? _dictionaryKeyConverters;
+
+        private static ConcurrentDictionary<Type, JsonConverter> GetDictionaryKeyConverters()
+        {
+            const int NumberOfConverters = 18;
+            var converters = new ConcurrentDictionary<Type, JsonConverter>(Environment.ProcessorCount, NumberOfConverters);
```

Comment:
nit: if we keep this cache, use the parameterless ConcurrentDictionary ctor here? See https://github.com/dotnet/runtime/pull/36726#discussion_r432201373 for a related discussion.

Reply:
The parameterless ctor internally uses the following.
```cs
private static int DefaultConcurrencyLevel => Environment.ProcessorCount;
```

This would achieve the same with the bonus of allowing us define a initial capacity to avoid resizing over and over. 
The difference with https://github.com/dotnet/runtime/pull/36726#discussion_r432201373 is that here, the capacity will only grow in the `Enum` case, which I don't think is very common.